packages/framework/ai-collab/src/explicit-strategy/index.ts

```diff
@@ -0,0 +1,362 @@
+		);
+		DEBUG_LOG?.push(planningPromt);
+		plan = await getStringFromLlm(planningPromt, options.openAI);
+	}
+
+	const originalDecoratedJson =
+		(options.finalReviewStep ?? false)
+			? toDecoratedJson(idGenerator, options.treeNode)
+			: undefined;
+	// reviewed is implicitly true if finalReviewStep is false
+	let hasReviewed = (options.finalReviewStep ?? false) ? false : true;
+	async function getNextEdit(): Promise<TreeEdit | undefined> {
+		const systemPrompt = getEditingSystemPrompt(
+			options.prompt.userAsk,
+			idGenerator,
+			options.treeView,
+			options.treeNode,
+			editLog,
+			options.prompt.systemRoleContext,
+			plan,
+		);
+
+		DEBUG_LOG?.push(systemPrompt);
+
+		const schema = types[rootTypeName] ?? fail("Root type not found.");
+		const wrapper = await getStructuredOutputFromLlm<EditWrapper>(
+			systemPrompt,
+			options.openAI,
+			schema,
+			"A JSON object that represents an edit to a JSON tree.",
+		);
+
+		// eslint-disable-next-line unicorn/no-null
+		DEBUG_LOG?.push(JSON.stringify(wrapper, null, 2));
+		if (wrapper === undefined) {
+			DEBUG_LOG?.push("Failed to get response");
+			return undefined;
+		}
+
+		if (wrapper.edit === null) {
+			DEBUG_LOG?.push("No more edits.");
+			if ((options.finalReviewStep ?? false) && !hasReviewed) {
+				const reviewResult = await reviewGoal();
+				if (reviewResult === undefined) {
+					DEBUG_LOG?.push("Failed to get review response");
+					return undefined;
+				}
+				// eslint-disable-next-line require-atomic-updates
+				hasReviewed = true;
+				if (reviewResult.goalAccomplished === "yes") {
+					return undefined;
+				} else {
+					// eslint-disable-next-line require-atomic-updates
+					editLog.length = 0;
+					return getNextEdit();
+				}
+			}
+		} else {
+			return wrapper.edit;
+		}
+	}
+
+	async function reviewGoal(): Promise<ReviewResult | undefined> {
+		const systemPrompt = getReviewSystemPrompt(
+			options.prompt.userAsk,
+			idGenerator,
+			options.treeView,
+			options.treeNode,
+			originalDecoratedJson ?? fail("Original decorated tree not provided."),
+			options.prompt.systemRoleContext,
+		);
+
+		DEBUG_LOG?.push(systemPrompt);
+
+		const schema = z.object({
+			goalAccomplished: z
+				.enum(["yes", "no"])
+				.describe('Whether the user\'s goal was met in the "after" tree.'),
+		});
+		return getStructuredOutputFromLlm<ReviewResult>(systemPrompt, options.openAI, schema);
+	}
+
+	let edit = await getNextEdit();
+	while (edit !== undefined) {
+		yield edit;
+		if (tokenUsage.inputTokens > (tokenLimits?.inputTokens ?? Number.POSITIVE_INFINITY)) {
+			throw new TokenLimitExceededError("Input token limit exceeded.");
+		}
+		if (tokenUsage.outputTokens > (tokenLimits?.outputTokens ?? Number.POSITIVE_INFINITY)) {
+			throw new TokenLimitExceededError("Output token limit exceeded.");
+		}
+		edit = await getNextEdit();
+	}
+}
+
+/**
+ * Calls an LLM to generate a structured output response based on the provided prompt.
```

Comment:
Nit
```suggestion
 * Calls the LLM to generate a structured output response based on the provided prompt.
```

Reply:
Updated in commit 6eb662389384cc2360c519307d3b33d3a7a02309